src/ast/passes/codegen_llvm.cpp

```diff
@@ -3771,28 +3771,41 @@ void CodegenLLVM::generate_maps(const RequiredResources &required_resources,
                       1,
                       MapKey({ CreateInt(loss_cnt_key_size) }),
                       CreateInt(loss_cnt_val_size));
-
-  if (required_resources.max_fmtstring_args_size > 0) {
-    createMapDefinition(to_string(MapType::FmtStringArgs),
-                        libbpf::BPF_MAP_TYPE_PERCPU_ARRAY,
-                        1,
-                        MapKey({ CreateInt32() }),
-                        CreateArray(required_resources.max_fmtstring_args_size,
-                                    CreateInt8()));
-  }
 }
 
 void CodegenLLVM::generate_global_vars(const RequiredResources &resources)
 {
   for (const auto &name : resources.needed_global_vars) {
-    auto var = llvm::dyn_cast<GlobalVariable>(
-        module_->getOrInsertGlobal(name, b_.getInt64Ty()));
-    var->setInitializer(b_.getInt64(1));
-    var->setConstant(true);
-    var->setSection(bpftrace::globalvars::SECTION_NAME);
-    var->setExternallyInitialized(true);
-    var->setDSOLocal(true);
-    var->addDebugInfo(debug_.createGlobalInt64(name));
+    if (name == globalvars::NUM_CPUS) {
+      auto var = llvm::dyn_cast<GlobalVariable>(
+          module_->getOrInsertGlobal(name, b_.getInt64Ty()));
+      var->setInitializer(b_.getInt64(1));
+      var->setConstant(true);
+      var->setSection(bpftrace::globalvars::RO_SECTION_NAME);
+      var->setExternallyInitialized(true);
+      var->setDSOLocal(true);
+      var->addDebugInfo(debug_.createGlobalInt64(name));
+    } else if (name == globalvars::FMT_STRINGS_BUFFER) {
+      assert(resources.max_fmtstring_args_size > 0);
+      // Assume 1 CPU to ensure codegen tests pass across multiple machines and
+      // so LLVM doesn't incorrectly optimize code (which happens if array size
+      // = 1)
+      auto array_type = CreateArray(resources.max_fmtstring_args_size,
+                                    CreateInt8());
+      auto var = llvm::dyn_cast<GlobalVariable>(
+          module_->getOrInsertGlobal(globalvars::FMT_STRINGS_BUFFER,
+                                     b_.GetType(array_type)));
+      var->setInitializer(ConstantAggregateZero::get(b_.GetType(array_type)));
+      var->setConstant(false);
+      var->setSection(bpftrace::globalvars::FMT_STRINGS_BUFFER_SECTION_NAME);
+      var->setExternallyInitialized(true);
+      var->setDSOLocal(true);
+      var->addDebugInfo(
+          debug_.createGlobalVariable(globalvars::FMT_STRINGS_BUFFER,
+                                      array_type));
+    } else {
+      LOG(BUG) << "Unknown global variable: " << name;
+    }
```

Comment:
Nit: maybe time for an enum? That way if we switch/case all the variants (leaving out default), compiler will warn if we forget to account for a new variant

Reply:
agreed, im gonna refactor globalvars stuff into enum and more configuration-driven